cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -524,26 +525,20 @@ private predicate getFieldSizeOfClass(Class c, Type type, int size) {
   )
 }
 
-private predicate simpleOperandLocalFlowStep(Operand opFrom, Instruction iTo) {
-  // Certain dataflow steps (for instance `PostUpdateNode.getPreUpdateNode()`) generates flow to
-  // operands, so we include dataflow from those operands to the "result" of the instruction (i.e., to
-  // the instruction itself).
-  exists(PostUpdateNode post |
-    opFrom = post.getPreUpdateNode().asOperand() and
-    iTo.getAnOperand() = opFrom
-  )
+private predicate simpleOperandLocalFlowStep(Instruction iFrom, Operand opTo) {
+  opTo.getAnyDef() = iFrom
```

Comment:
I don't think we should allow all flow from instructions to operands. That leads to requiring side conditions of `not opFrom.isDefinitionInexact()` in the operand-to-instruction modelling below, but these conditions are not about the operand-to-instruction step. They're really about the preceding-instruction-to-operand step. Users should not have to write `not opFrom.isDefinitionInexact()` when adding custom steps.

The way I envision these rules, they're split into
- **Operand-to-instruction flow.** These rules are determined by the semantics of the opcode. For example, `iTo.(CopyInstruction).getSourceValueOperand() = opFrom` would be such a rule, without any further conditions. Once flow has reached the operand of an instruction, we should no longer think about how it got there. Control-flow-wise, I think of an `Operand`s as being evaluated right before their `Instruction` because that's where we want them for the purpose of guards.
- **Instruction-to-operand flow.** These rules determine whether an operand should get flow from the instruction that defined its value. Overlap is key here, and the main rule should be `opTo.getDef() = iFrom`. Those of our rules that currently break this principle and require more flow than `getDef` allows are exactly all the dodgy ones, and I'm fine with adding additional rules for those specifically.

The rule for `ReadSideEffectInstruction` is special and belongs in instruction-to-operand flow rather than operand-to-instruction flow. That's because flow to an instruction means flow _out of_ the instruction, but there isn't supposed to be flow out of this instruction. With the changes you're making here, it's finally possible to address https://github.com/github/codeql/pull/3123#discussion_r433916902. Note that this requires corresponding changes over in `DataFlowPrivate.qll`.

If it's convenient or performant, the rules for instruction-to-operand flow can be a unary predicate on `Operand` because for any given `Operand` it's always clear which instruction the flow comes from.